src/binaryServerStarter.ts

```diff
@@ -0,0 +1,285 @@
+import * as os from 'os';
+import * as path from 'path';
+import { Readable } from 'stream';
+import { commands, ExtensionContext, extensions, window, WorkspaceConfiguration } from "vscode";
+import { Executable } from "vscode-languageclient";
+import * as yauzl from 'yauzl';
+import { addTrustedHash, getTrustedHashes } from './binaryHashManager';
+import { getXMLConfiguration } from "./settings";
+const glob = require('glob');
+
+const HTTPS_PATTERN: RegExp = /^https:\/\//;
+const JAR_ZIP_AND_HASH_REJECTOR: RegExp = /(?:\.jar)|(?:\.zip)|(?:\.sha256)$/;
+
+/**
+ * Returns the executable to launch LemMinX (the XML Language Server) as a binary
+ *
+ * @param context the extension context
+ * @throws if the binary doesn't exist and can't be downloaded, or if the binary is not trusted
+ * @returns Returns the executable to launch LemMinX (the XML Language Server) as a binary
+ */
+export async function prepareBinaryExecutable(context: ExtensionContext): Promise<Executable> {
+  const binaryOptions: string = getXMLConfiguration().get("server.binary.args");
+  let binaryExecutable: Executable;
+  return getServerBinaryPath()
+    .then((binaryPath: string) => {
+      binaryExecutable = {
+        args: [binaryOptions],
+        command: binaryPath
+      } as Executable;
+      return binaryPath;
+    })
+    .then(checkBinaryHash)
+    .then((hashOk: boolean) => {
+      if (hashOk) {
+        return binaryExecutable;
+      } else {
+        throw new Error("The binary XML language server is not trusted");
+      }
+    });
+}
+
+/**
+ * Returns the path to the LemMinX binary
+ *
+ * Downloads it if it is missing
+ *
+ * @returns The path to the LemMinX binary
+ * @throws If the LemMinX binary can't be located or downloaded
+ */
+async function getServerBinaryPath(): Promise<string> {
+  const config: WorkspaceConfiguration = getXMLConfiguration();
+  let binaryPath: string = config.get("server.binary.path");
+  if (binaryPath) {
+    if (fs.existsSync(binaryPath)) {
+      return Promise.resolve(binaryPath);
+    }
+    window.showErrorMessage('The specified XML language server binary could not be found. Using the default binary...');
+  }
+  let server_home: string = path.resolve(__dirname, '../server');
+  let binaries: Array<string> = glob.sync(`**/${getServerBinaryNameWithoutExtension()}*`, { cwd: server_home });
+  binaries = binaries.filter((path) => { return !JAR_ZIP_AND_HASH_REJECTOR.test(path) });
+  if (binaries.length) {
+    return Promise.resolve(path.resolve(server_home, binaries[0]));
+  }
+  // Download it, then return the downloaded binary's location
+  return downloadBinary();
+}
+
+/**
+ * Downloads LemMinX binary under the `server` directory and returns the path to the binary as a Promise
+ *
+ * @returns The path to the LemMinX binary
+ * @throws If the LemMinX binary download fails
+ */
+async function downloadBinary(): Promise<string> {
+  window.setStatusBarMessage('Downloading XML language server binary...', 2000);
```

Comment:
can you use `setStatusBarMessage(text: string, hideWhenDone: Thenable<any>)` instead?
I was thrown off by seeing the message disappear quickly and still waiting 10 sec before seeing any XML activity